[AUTOCUT] Gradle Check Flaky Test Report for FullRollingRestartIT

## Flaky Test Report for `FullRollingRestartIT`

Noticed the `FullRollingRestartIT` has some flaky, failing tests that failed during **post-merge actions** and **timer trigerred run on main branch**.

### Details

| Git Reference | Merged Pull Request | Build Details | Test Name |
|---------------|----------------------|---------------|-----------|
| main | [null](https://github.com/opensearch-project/OpenSearch/pull/null) | [60843](https://build.ci.opensearch.org/job/gradle-check/60843/testReport/) <br><br>[60063](https://build.ci.opensearch.org/job/gradle-check/60063/testReport/) <br><br>[60317](https://build.ci.opensearch.org/job/gradle-check/60317/testReport/) <br><br>[60324](https://build.ci.opensearch.org/job/gradle-check/60324/testReport/) <br><br>[60337](https://build.ci.opensearch.org/job/gradle-check/60337/testReport/) <br><br>[60345](https://build.ci.opensearch.org/job/gradle-check/60345/testReport/) <br><br>[60450](https://build.ci.opensearch.org/job/gradle-check/60450/testReport/) <br><br>[60482](https://build.ci.opensearch.org/job/gradle-check/60482/testReport/) <br><br>[60567](https://build.ci.opensearch.org/job/gradle-check/60567/testReport/) <br><br>[60733](https://build.ci.opensearch.org/job/gradle-check/60733/testReport/) <br><br>[60753](https://build.ci.opensearch.org/job/gradle-check/60753/testReport/) <br><br>[60763](https://build.ci.opensearch.org/job/gradle-check/60763/testReport/) <br><br>[60793](https://build.ci.opensearch.org/job/gradle-check/60793/testReport/) <br><br>[60807](https://build.ci.opensearch.org/job/gradle-check/60807/testReport/) <br><br>[60831](https://build.ci.opensearch.org/job/gradle-check/60831/testReport/) <br><br>[60890](https://build.ci.opensearch.org/job/gradle-check/60890/testReport/) <br><br>[60925](https://build.ci.opensearch.org/job/gradle-check/60925/testReport/) <br><br>[60949](https://build.ci.opensearch.org/job/gradle-check/60949/testReport/) <br><br>[60982](https://build.ci.opensearch.org/job/gradle-check/60982/testReport/) <br><br>[61009](https://build.ci.opensearch.org/job/gradle-check/61009/testReport/) <br><br>[61012](https://build.ci.opensearch.org/job/gradle-check/61012/testReport/) <br><br>[61022](https://build.ci.opensearch.org/job/gradle-check/61022/testReport/) <br><br>[61075](https://build.ci.opensearch.org/job/gradle-check/61075/testReport/) <br><br>[61082](https://build.ci.opensearch.org/job/gradle-check/61082/testReport/) <br><br>[61090](https://build.ci.opensearch.org/job/gradle-check/61090/testReport/) <br><br>[61095](https://build.ci.opensearch.org/job/gradle-check/61095/testReport/) <br><br>[61098](https://build.ci.opensearch.org/job/gradle-check/61098/testReport/) <br><br>[61261](https://build.ci.opensearch.org/job/gradle-check/61261/testReport/) <br><br>[61276](https://build.ci.opensearch.org/job/gradle-check/61276/testReport/) <br><br>[61322](https://build.ci.opensearch.org/job/gradle-check/61322/testReport/) <br><br>[61331](https://build.ci.opensearch.org/job/gradle-check/61331/testReport/) <br><br>[61426](https://build.ci.opensearch.org/job/gradle-check/61426/testReport/) <br><br>[61528](https://build.ci.opensearch.org/job/gradle-check/61528/testReport/) <br><br>[61529](https://build.ci.opensearch.org/job/gradle-check/61529/testReport/) <br><br>[61553](https://build.ci.opensearch.org/job/gradle-check/61553/testReport/) <br><br>[61623](https://build.ci.opensearch.org/job/gradle-check/61623/testReport/) <br><br>[61644](https://build.ci.opensearch.org/job/gradle-check/61644/testReport/) <br><br>[61689](https://build.ci.opensearch.org/job/gradle-check/61689/testReport/) <br><br>[61759](https://build.ci.opensearch.org/job/gradle-check/61759/testReport/) <br><br>[61830](https://build.ci.opensearch.org/job/gradle-check/61830/testReport/) <br><br>[61838](https://build.ci.opensearch.org/job/gradle-check/61838/testReport/) <br><br>[61892](https://build.ci.opensearch.org/job/gradle-check/61892/testReport/) <br><br>[71099](https://build.ci.opensearch.org/job/gradle-check/71099/testReport/) <br><br>[71188](https://build.ci.opensearch.org/job/gradle-check/71188/testReport/) <br><br>[71259](https://build.ci.opensearch.org/job/gradle-check/71259/testReport/) <br><br>[71351](https://build.ci.opensearch.org/job/gradle-check/71351/testReport/) <br><br>[71403](https://build.ci.opensearch.org/job/gradle-check/71403/testReport/) <br><br>[71409](https://build.ci.opensearch.org/job/gradle-check/71409/testReport/) | `org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart {p0={"cluster.indices.replication.strategy":"SEGMENT"}}`<br><br>`org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart_withNoRecoveryPayloadAndSource {p0={"cluster.indices.replication.strategy":"SEGMENT"}}`<br><br>`org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart {p0={"cluster.indices.replication.strategy":"DOCUMENT"}}` |
| c94203d36b71c2c7f033031a6afae935c3b42d8a | [20587](https://github.com/opensearch-project/OpenSearch/pull/20587) | [71279](https://build.ci.opensearch.org/job/gradle-check/71279/testReport/) | `org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart_withNoRecoveryPayloadAndSource {p0={"cluster.indices.replication.strategy":"SEGMENT"}}` |
| e0e7f2935a1d6990fe2f2741328711bdb71f8e84 | [20589](https://github.com/opensearch-project/OpenSearch/pull/20589) | [71205](https://build.ci.opensearch.org/job/gradle-check/71205/testReport/) | `org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart_withNoRecoveryPayloadAndSource {p0={"cluster.indices.replication.strategy":"SEGMENT"}}` |
| 09b6a36ea98862ea7c13d530f619835d43d3c087 | [20551](https://github.com/opensearch-project/OpenSearch/pull/20551) | [71085](https://build.ci.opensearch.org/job/gradle-check/71085/testReport/) | `org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart_withNoRecoveryPayloadAndSource {p0={"cluster.indices.replication.strategy":"SEGMENT"}}` |
| c5d26f7f637f38f6e9aed2553de47a9b2ad244fa | [18857](https://github.com/opensearch-project/OpenSearch/pull/18857) | [61858](https://build.ci.opensearch.org/job/gradle-check/61858/testReport/) | `org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart_withNoRecoveryPayloadAndSource {p0={"cluster.indices.replication.strategy":"SEGMENT"}}` |
| b55b140ceab31669e20b21c3dc99f4a7a70d1cf9 | [18904](https://github.com/opensearch-project/OpenSearch/pull/18904) | [61695](https://build.ci.opensearch.org/job/gradle-check/61695/testReport/) | `org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart {p0={"cluster.indices.replication.strategy":"SEGMENT"}}` |
| b0417f357b7a2cea17297da14b2ce619a9ca2475 | [18511](https://github.com/opensearch-project/OpenSearch/pull/18511) | [61205](https://build.ci.opensearch.org/job/gradle-check/61205/testReport/) | `org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart {p0={"cluster.indices.replication.strategy":"SEGMENT"}}` |
| 36844f5d5b3ad4ab4d35f7855401c2c1d27210dd | [18754](https://github.com/opensearch-project/OpenSearch/pull/18754) | [61117](https://build.ci.opensearch.org/job/gradle-check/61117/testReport/) | `org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart {p0={"cluster.indices.replication.strategy":"SEGMENT"}}` |
| c4d9de6180800844afcd9ab16f93082d557773a9 | [18767](https://github.com/opensearch-project/OpenSearch/pull/18767) | [60884](https://build.ci.opensearch.org/job/gradle-check/60884/testReport/) | `org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart {p0={"cluster.indices.replication.strategy":"SEGMENT"}}` |
| 6119faa7a0040b188246ffd4de6ea2b24efcb398 | [18749](https://github.com/opensearch-project/OpenSearch/pull/18749) | [60845](https://build.ci.opensearch.org/job/gradle-check/60845/testReport/) | `org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart {p0={"cluster.indices.replication.strategy":"SEGMENT"}}` |
| 135262712bae5f0ac77fc2f27eea50e7d4a912aa | [18707](https://github.com/opensearch-project/OpenSearch/pull/18707) | [60548](https://build.ci.opensearch.org/job/gradle-check/60548/testReport/) | `org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart {p0={"cluster.indices.replication.strategy":"SEGMENT"}}` |
| cb22cb02c9f9a4acd40f4c460ce9247a4c5b127b | [18661](https://github.com/opensearch-project/OpenSearch/pull/18661) | [60520](https://build.ci.opensearch.org/job/gradle-check/60520/testReport/) | `org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart_withNoRecoveryPayloadAndSource {p0={"cluster.indices.replication.strategy":"SEGMENT"}}` |
| 624f85a9fcf63006517907c758482ef164bd228d | [18743](https://github.com/opensearch-project/OpenSearch/pull/18743) | [60468](https://build.ci.opensearch.org/job/gradle-check/60468/testReport/) | `org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart {p0={"cluster.indices.replication.strategy":"SEGMENT"}}` |
| f7ebe22b39f8c2ccae6f22f9b7776995a2c19be1 | [18488](https://github.com/opensearch-project/OpenSearch/pull/18488) | [59157](https://build.ci.opensearch.org/job/gradle-check/59157/testReport/) | `org.opensearch.recovery.FullRollingRestartIT.classMethod`<br><br>`org.opensearch.recovery.FullRollingRestartIT.testDerivedSourceWithConcurrentUpdatesRollingRestart {p0={"cluster.indices.replication.strategy":"SEGMENT"}}` |
| 8a59ca1d417cb6022531125df09e274997c27ef3 | [18277](https://github.com/opensearch-project/OpenSearch/pull/18277) | [59141](https://build.ci.opensearch.org/job/gradle-check/59141/testReport/) | `org.opensearch.recovery.FullRollingRestartIT.classMethod`<br><br>`org.opensearch.recovery.FullRollingRestartIT.testDerivedSourceWithConcurrentUpdatesRollingRestart {p0={"cluster.indices.replication.strategy":"DOCUMENT"}}` |
| c92b8ea8742b7ae48e5b169c02a255543c4c7b5d | [18435](https://github.com/opensearch-project/OpenSearch/pull/18435) | [59128](https://build.ci.opensearch.org/job/gradle-check/59128/testReport/) | `org.opensearch.recovery.FullRollingRestartIT.testDerivedSourceWithMixedVersionRollingRestart {p0={"cluster.indices.replication.strategy":"DOCUMENT"}}` |
| d0c2e39ae05454775b8063e09a88dd5f5834c49f | [17797](https://github.com/opensearch-project/OpenSearch/pull/17797) | [59086](https://build.ci.opensearch.org/job/gradle-check/59086/testReport/) | `org.opensearch.recovery.FullRollingRestartIT.classMethod`<br><br>`org.opensearch.recovery.FullRollingRestartIT.testDerivedSourceWithConcurrentUpdatesRollingRestart {p0={"cluster.indices.replication.strategy":"SEGMENT"}}` |

The other pull requests, besides those involved in post-merge actions, that contain failing tests with the `FullRollingRestartIT` class are:

- [18504](https://github.com/opensearch-project/OpenSearch/pull/18504)
- [18479](https://github.com/opensearch-project/OpenSearch/pull/18479)
- [18467](https://github.com/opensearch-project/OpenSearch/pull/18467)
- [18497](https://github.com/opensearch-project/OpenSearch/pull/18497)
- [18517](https://github.com/opensearch-project/OpenSearch/pull/18517)
- [18070](https://github.com/opensearch-project/OpenSearch/pull/18070)
- [18195](https://github.com/opensearch-project/OpenSearch/pull/18195)
- [18229](https://github.com/opensearch-project/OpenSearch/pull/18229)
- [18351](https://github.com/opensearch-project/OpenSearch/pull/18351)
- [18375](https://github.com/opensearch-project/OpenSearch/pull/18375)
- [18664](https://github.com/opensearch-project/OpenSearch/pull/18664)
- [18685](https://github.com/opensearch-project/OpenSearch/pull/18685)
- [18887](https://github.com/opensearch-project/OpenSearch/pull/18887)
- [18896](https://github.com/opensearch-project/OpenSearch/pull/18896)
- [18672](https://github.com/opensearch-project/OpenSearch/pull/18672)
- [18702](https://github.com/opensearch-project/OpenSearch/pull/18702)
- [18724](https://github.com/opensearch-project/OpenSearch/pull/18724)
- [18805](https://github.com/opensearch-project/OpenSearch/pull/18805)
- [18936](https://github.com/opensearch-project/OpenSearch/pull/18936)
- [18333](https://github.com/opensearch-project/OpenSearch/pull/18333)
- [18424](https://github.com/opensearch-project/OpenSearch/pull/18424)
- [18451](https://github.com/opensearch-project/OpenSearch/pull/18451)
- [18565](https://github.com/opensearch-project/OpenSearch/pull/18565)
- [18656](https://github.com/opensearch-project/OpenSearch/pull/18656)
- [18706](https://github.com/opensearch-project/OpenSearch/pull/18706)
- [18777](https://github.com/opensearch-project/OpenSearch/pull/18777)
- [18800](https://github.com/opensearch-project/OpenSearch/pull/18800)
- [18877](https://github.com/opensearch-project/OpenSearch/pull/18877)
- [18890](https://github.com/opensearch-project/OpenSearch/pull/18890)
- [16291](https://github.com/opensearch-project/OpenSearch/pull/16291)
- [17741](https://github.com/opensearch-project/OpenSearch/pull/17741)
- [18066](https://github.com/opensearch-project/OpenSearch/pull/18066)
- [18491](https://github.com/opensearch-project/OpenSearch/pull/18491)
- [18531](https://github.com/opensearch-project/OpenSearch/pull/18531)
- [18550](https://github.com/opensearch-project/OpenSearch/pull/18550)
- [18568](https://github.com/opensearch-project/OpenSearch/pull/18568)
- [18639](https://github.com/opensearch-project/OpenSearch/pull/18639)
- [18668](https://github.com/opensearch-project/OpenSearch/pull/18668)
- [18671](https://github.com/opensearch-project/OpenSearch/pull/18671)
- [18691](https://github.com/opensearch-project/OpenSearch/pull/18691)
- [18697](https://github.com/opensearch-project/OpenSearch/pull/18697)
- [18712](https://github.com/opensearch-project/OpenSearch/pull/18712)
- [18720](https://github.com/opensearch-project/OpenSearch/pull/18720)
- [18732](https://github.com/opensearch-project/OpenSearch/pull/18732)
- [18735](https://github.com/opensearch-project/OpenSearch/pull/18735)
- [18741](https://github.com/opensearch-project/OpenSearch/pull/18741)
- [18746](https://github.com/opensearch-project/OpenSearch/pull/18746)
- [18748](https://github.com/opensearch-project/OpenSearch/pull/18748)
- [18752](https://github.com/opensearch-project/OpenSearch/pull/18752)
- [18766](https://github.com/opensearch-project/OpenSearch/pull/18766)
- [18779](https://github.com/opensearch-project/OpenSearch/pull/18779)
- [18803](https://github.com/opensearch-project/OpenSearch/pull/18803)
- [18814](https://github.com/opensearch-project/OpenSearch/pull/18814)
- [18825](https://github.com/opensearch-project/OpenSearch/pull/18825)
- [18837](https://github.com/opensearch-project/OpenSearch/pull/18837)
- [18842](https://github.com/opensearch-project/OpenSearch/pull/18842)
- [18843](https://github.com/opensearch-project/OpenSearch/pull/18843)
- [18844](https://github.com/opensearch-project/OpenSearch/pull/18844)
- [18874](https://github.com/opensearch-project/OpenSearch/pull/18874)
- [18885](https://github.com/opensearch-project/OpenSearch/pull/18885)
- [18889](https://github.com/opensearch-project/OpenSearch/pull/18889)
- [18901](https://github.com/opensearch-project/OpenSearch/pull/18901)
- [18914](https://github.com/opensearch-project/OpenSearch/pull/18914)
- [18929](https://github.com/opensearch-project/OpenSearch/pull/18929)
- [18933](https://github.com/opensearch-project/OpenSearch/pull/18933)
- [20150](https://github.com/opensearch-project/OpenSearch/pull/20150)
- [20503](https://github.com/opensearch-project/OpenSearch/pull/20503)
- [20558](https://github.com/opensearch-project/OpenSearch/pull/20558)
- [20566](https://github.com/opensearch-project/OpenSearch/pull/20566)
- [20585](https://github.com/opensearch-project/OpenSearch/pull/20585)
- [20607](https://github.com/opensearch-project/OpenSearch/pull/20607)
- [20623](https://github.com/opensearch-project/OpenSearch/pull/20623)
- [20632](https://github.com/opensearch-project/OpenSearch/pull/20632)
- [20648](https://github.com/opensearch-project/OpenSearch/pull/20648)

For more details on the failed tests refer to [OpenSearch Gradle Check Metrics](https://metrics.opensearch.org/_dashboards/app/dashboards#/view/e5e64d40-ed31-11ee-be99-69d1dbc75083) dashboard.

<img width="1269" height="420" alt="Image" src="https://github.com/user-attachments/assets/ea8dc6e2-a86f-4202-a34a-5a4c456d514e" />

@ashking94 The `testFullRollingRestart` and `testFullRollingRestart_withNoRecoveryPayloadAndSource` test cases started failing around July 6 with [no recorded failures in the previous year](https://metrics.opensearch.org/_dashboards/app/dashboards#/view/e5e64d40-ed31-11ee-be99-69d1dbc75083?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-12M,to:now))&_a=(description:'',filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:'2e729590-eb26-11ee-b008-6583d3dd450c',key:invoke_type.keyword,negate:!f,params:!(Timer,'Post%20Merge%20Action'),type:phrases,value:'Timer,%20Post%20Merge%20Action'),query:(bool:(minimum_should_match:1,should:!((match_phrase:(invoke_type.keyword:Timer)),(match_phrase:(invoke_type.keyword:'Post%20Merge%20Action')))))),('$state':(store:appState),meta:(alias:!n,disabled:!f,index:'2e729590-eb26-11ee-b008-6583d3dd450c',key:test_class,negate:!f,params:!(FullRollingRestartIT),type:phrases,value:FullRollingRestartIT),query:(bool:(minimum_should_match:1,should:!((match_phrase:(test_class:FullRollingRestartIT)))))),('$state':(store:appState),meta:(alias:!n,disabled:!f,index:'2e729590-eb26-11ee-b008-6583d3dd450c',key:test_name,negate:!f,params:!('org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart%20%7Bp0%3D%7B%22cluster.indices.replication.strategy%22:%22SEGMENT%22%7D%7D','org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart_withNoRecoveryPayloadAndSource%20%7Bp0%3D%7B%22cluster.indices.replication.strategy%22:%22SEGMENT%22%7D%7D'),type:phrases,value:'org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart%20%7Bp0%3D%7B%22cluster.indices.replication.strategy%22:%22SEGMENT%22%7D%7D,%20org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart_withNoRecoveryPayloadAndSource%20%7Bp0%3D%7B%22cluster.indices.replication.strategy%22:%22SEGMENT%22%7D%7D'),query:(bool:(minimum_should_match:1,should:!((match_phrase:(test_name:'org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart%20%7Bp0%3D%7B%22cluster.indices.replication.strategy%22:%22SEGMENT%22%7D%7D')),(match_phrase:(test_name:'org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart_withNoRecoveryPayloadAndSource%20%7Bp0%3D%7B%22cluster.indices.replication.strategy%22:%22SEGMENT%22%7D%7D'))))))),fullScreenMode:!f,options:(hidePanelTitles:!f,useMargins:!t),query:(language:kuery,query:''),timeRestore:!t,title:'OpenSearch%20Gradle%20Check%20Report',viewMode:view)). It only happens with segment replication. I am unable to reproduce it locally unfortunately. Every failure I've seen is caused by the test timing out while doing a `refreshAndWaitForReplication()` call (example failure below). There was a change to segrep from commit 46a0045a2301b3bc51e2da9c5e853f69ac6342c2 on July 3, which is just a few days before the first observed failure. Is there any chance that commit introduced a bug in this rolling restart scenario?

```
REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests 'org.opensearch.recovery.FullRollingRestartIT' -Dtests.method='testFullRollingRestart_withNoRecoveryPayloadAndSource {p0={"cluster.indices.replication.strategy":"SEGMENT"}}' -Dtests.seed=961918EE833A9E37 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=fr-NC -Dtests.timezone=CTT -Druntime.java=24

FullRollingRestartIT > testFullRollingRestart_withNoRecoveryPayloadAndSource {p0={"cluster.indices.replication.strategy":"SEGMENT"}} FAILED
    java.lang.AssertionError: replica shards haven't caught up with primary expected:<20> but was:<17>
        at __randomizedtesting.SeedInfo.seed([961918EE833A9E37:A55072AE2A77FF6F]:0)
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:647)
        at org.opensearch.test.OpenSearchIntegTestCase.lambda$waitForReplication$1(OpenSearchIntegTestCase.java:2550)
        at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1181)
        at org.opensearch.test.OpenSearchIntegTestCase.waitForReplication(OpenSearchIntegTestCase.java:2539)
        at org.opensearch.test.OpenSearchIntegTestCase.refreshAndWaitForReplication(OpenSearchIntegTestCase.java:2512)
        at org.opensearch.recovery.FullRollingRestartIT.testFullRollingRestart_withNoRecoveryPayloadAndSource(FullRollingRestartIT.java:360)
```
https://build.ci.opensearch.org/job/gradle-check/61468/consoleText

--- a/server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java
+++ b/server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java
@@ -364,10 +364,18 @@ public class SegmentReplicationTargetService extends AbstractLifecycleComponent 
                         // update visible checkpoint to primary
                         updateVisibleCheckpoint(state.getReplicationId(), replicaShard);
 
-                        // if we received a checkpoint during the copy event that is ahead of this
-                        // try and process it.
+                        // If the latest checkpoint known from the primary is still ahead of what this replica has
+                        // actually replicated, process it to catch up. We compare against the replica's achieved
+                        // checkpoint rather than the checkpoint this round targeted (receivedCheckpoint). A retry
+                        // round may finalize against a stale metadata checkpoint returned by the primary (see #20550
+                        // and #20551), leaving the replica behind the checkpoint it was asked to sync to. Comparing
+                        // against the targeted checkpoint would miss this case and leave the replica permanently
+                        // behind until the next publish; comparing against the achieved checkpoint re-triggers the
+                        // catch-up. In the normal case the achieved checkpoint is at least the targeted one, so this
+                        // does not introduce extra rounds.
                         ReplicationCheckpoint latestReceivedCheckpoint = replicator.getPrimaryCheckpoint(replicaShard.shardId());
-                        if (Objects.nonNull(latestReceivedCheckpoint) && latestReceivedCheckpoint.isAheadOf(receivedCheckpoint)) {
+                        if (Objects.nonNull(latestReceivedCheckpoint)
+                            && latestReceivedCheckpoint.isAheadOf(replicaShard.getLatestReplicationCheckpoint())) {
                             processLatestReceivedCheckpoint(replicaShard, thread);
                         }
                     }

--- a/server/src/test/java/org/opensearch/indices/replication/SegmentReplicationTargetServiceTests.java
+++ b/server/src/test/java/org/opensearch/indices/replication/SegmentReplicationTargetServiceTests.java
@@ -66,6 +66,8 @@ import java.util.Set;
 import java.util.UUID;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.BiConsumer;
 
 import static org.opensearch.index.seqno.SequenceNumbers.NO_OPS_PERFORMED;
@@ -630,18 +632,63 @@ public class SegmentReplicationTargetServiceTests extends IndexShardTestCase {
     public void testStartReplicationListenerSuccess() throws InterruptedException {
         sut.updateLatestReceivedCheckpoint(aheadCheckpoint, replicaShard);
         SegmentReplicationTargetService spy = spy(sut);
+        // Model the replica's achieved checkpoint. A successful round advances it to the checkpoint it targeted, so
+        // the done-handler observes the replica has caught up and does not retrigger another round.
+        IndexShard spyReplicaShard = spy(replicaShard);
+        AtomicReference<ReplicationCheckpoint> achieved = new AtomicReference<>(replicaShard.getLatestReplicationCheckpoint());
+        doAnswer(i -> achieved.get()).when(spyReplicaShard).getLatestReplicationCheckpoint();
         CountDownLatch latch = new CountDownLatch(1);
         doAnswer(i -> {
+            achieved.set(i.getArgument(1));
             ((SegmentReplicationTargetService.SegmentReplicationListener) i.getArgument(3)).onReplicationDone(state);
             latch.countDown();
             return null;
         }).when(spy).startReplication(any(), any(), anyBoolean(), any());
         doNothing().when(spy).updateVisibleCheckpoint(eq(0L), any());
-        spy.afterIndexShardStarted(replicaShard);
+        spy.afterIndexShardStarted(spyReplicaShard);
 
         latch.await(2, TimeUnit.SECONDS);
-        verify(spy, (atLeastOnce())).updateVisibleCheckpoint(eq(0L), eq(replicaShard));
+        verify(spy, (atLeastOnce())).updateVisibleCheckpoint(eq(0L), eq(spyReplicaShard));
+        // Only the initial entry from afterIndexShardStarted; the done-handler must not retrigger since the replica
+        // caught up to the targeted checkpoint.
         verify(spy, times(1)).processLatestReceivedCheckpoint(any(), any());
+    }
+
+    /**
+     * Regression test for the scenario described in #20550 / #20551. After a failed round is retried, the retry may
+     * finalize against a stale metadata checkpoint returned by the primary, leaving the replica behind the checkpoint
+     * it was asked to sync to. The done-handler must detect that the replica is still behind the primary's latest
+     * known checkpoint - by comparing against the replica's achieved checkpoint, not the checkpoint the round targeted -
+     * and trigger another round so the replica eventually catches up. Here the first round finalizes without advancing
+     * the replica's checkpoint (stale sync); the second round advances it, after which no further round is triggered.
+     */
+    public void testStartReplicationRetriggersWhenReplicaSyncedToStaleCheckpoint() throws InterruptedException {
+        sut.updateLatestReceivedCheckpoint(aheadCheckpoint, replicaShard);
+        SegmentReplicationTargetService spy = spy(sut);
+        IndexShard spyReplicaShard = spy(replicaShard);
+        AtomicReference<ReplicationCheckpoint> achieved = new AtomicReference<>(replicaShard.getLatestReplicationCheckpoint());
+        doAnswer(i -> achieved.get()).when(spyReplicaShard).getLatestReplicationCheckpoint();
+        AtomicInteger rounds = new AtomicInteger(0);
+        CountDownLatch latch = new CountDownLatch(2);
+        doAnswer(i -> {
+            // First round finalizes against a stale checkpoint and does NOT advance the replica. The second
+            // (retriggered) round successfully advances the replica to the checkpoint it was asked to sync to.
+            if (rounds.incrementAndGet() >= 2) {
+                achieved.set(i.getArgument(1));
+            }
+            ((SegmentReplicationTargetService.SegmentReplicationListener) i.getArgument(3)).onReplicationDone(state);
+            latch.countDown();
+            return null;
+        }).when(spy).startReplication(any(), any(), anyBoolean(), any());
+        doNothing().when(spy).updateVisibleCheckpoint(eq(0L), any());
+        spy.afterIndexShardStarted(spyReplicaShard);
+
+        assertTrue("expected the replica to be retriggered until it caught up", latch.await(5, TimeUnit.SECONDS));
+        // Two rounds: the initial stale round plus the retriggered catch-up round.
+        verify(spy, times(2)).startReplication(eq(spyReplicaShard), any(), anyBoolean(), any());
+        // processLatestReceivedCheckpoint: once from afterIndexShardStarted and once from the done-handler after the
+        // first (stale) round. The second round catches up, so the done-handler does not retrigger again.
+        verify(spy, org.mockito.Mockito.atLeast(2)).processLatestReceivedCheckpoint(any(), any());
     }
 
     public void testStartReplicationListenerFailure() throws InterruptedException {

